### ISSUE TITLE
Question about using construct to parse partial packets

### ISSUE THREAD

**Issue by nathanaelhuffman** [construct/construct#389]
Ideally I'd like to define the whole packet structure but be able to pass in fewer bytes than required and parse what bytes are present into a packet structure.

I have various needs for full packet definition (packet generation, packet parsing, packet or protocol documentation generation etc) and would like to be able to use a single set of definitions of packets for the following use-case as well:

My use-case is I have data-dumps that have the first N bytes of a series of different packets, but depending on how the data capture is set up, N can vary capture to capture.  I'd like to define the whole packet structure and then take the bytes that I have and try to parse them into various types of defined packets. 

I'm ok to do some wrapping or extending of construct to make this work but some comments on whether this is possible or where I might start extending would be helpful

**Comment by arekbulski** [comment]
TLDR. Look at `Union` and `Select` classes.

**Comment by nathanaelhuffman** [comment]
Thanks for your quick response.  I played around with this a bit today and it looks like `Select` is part of the answer.  Making some fields `Optional` also helps but I still get un-desired results when the data doesn't end on a nice boundary (which is what I need for my use-case):

```python
from construct import *
import binascii

hdr = 'header' / Struct(
    'src' / Int8ub,
    'dest' / Int8ub,
    'type' / Const(Int8ub, 5),
    'count' / Int8ub,
)
packet = 'pkt' / Struct(
    'hdr' / hdr,
    'payload' / Optional(Int32ub),
    'checksum' / Optional(Int16ub)
)
print(Select(packet, hdr).parse(binascii.unhexlify('55440505DEADBEEFAA55')))  # Ok, full decode of packet
print(Select(packet, hdr).parse(binascii.unhexlify('55440505DEADBEEFAA')))      # Ok, packet decoded but checksum = None
print(Select(packet, hdr).parse(binascii.unhexlify('55440504DEADBE')))              # Odd, DEAD applied to checksum since payload was optional
```
```

**Comment by arekbulski** [comment]
You could write your own class that does something like that... a modified Union/Select but... you would have to understand how the library works throughly. It might be a rewarding experience, but would take you  a day or two to just learn how this thing operates. Just something for you to consider.

**Comment by arekbulski** [comment]
PS you can use Byte/Short/Int as aliases.

**Comment by nathanaelhuffman** [comment]
@arekbulski ok thanks.  I wanted to confirm that this would have to be extended functionality.  This is a low priority side-project for me but I'll think about digging in and see where it goes.  Thanks for your help!
